software/fpga/ov3/ovhw/whacker/producer.py

```diff
@@ -188,34 +208,63 @@ def write_hdr(statename, nextname, hdr_offs, val):
+                            NextState("WRT5")
+                        ).Elif(rel_timestamp[32:40],
+                            self.delta_ts_size.set(4),
+                            NextState("WRT4")
+                        ).Elif(rel_timestamp[24:32],
+                            self.delta_ts_size.set(3),
+                            NextState("WRT3")
+                        ).Elif(rel_timestamp[16:24],
+                            self.delta_ts_size.set(2),
+                            NextState("WRT2")
+                        ).Elif(rel_timestamp[8:16],
+                            self.delta_ts_size.set(1),
+                            NextState("WRT1")
+                        ).Else(
+                            self.delta_ts_size.set(0),
+                            NextState("WRT0")
+                        ),
+                        clear_acc_flags.eq(1))
                 ))
 
-        write_hdr("WH0", "WRF0", 0, 0xA0)
-
-        # Write flags field
-        write_hdr("WRF0", "WRF1", 1, self.flags.v[:8])
-        write_hdr("WRF1", "WRSL", 2, self.flags.v[8:16])
+        # Write header beginning with last header byte
+        write_hdr("WRT7", "WRT6", self.delta_timestamp.v[56:64])
+        write_hdr("WRT6", "WRT5", self.delta_timestamp.v[48:56])
+        write_hdr("WRT5", "WRT4", self.delta_timestamp.v[40:48])
+        write_hdr("WRT4", "WRT3", self.delta_timestamp.v[32:40])
+        write_hdr("WRT3", "WRT2", self.delta_timestamp.v[24:32])
+        write_hdr("WRT2", "WRT1", self.delta_timestamp.v[16:24])
+        write_hdr("WRT1", "WRT0", self.delta_timestamp.v[8:16])
+        write_hdr("WRT0", "WRSH", self.delta_timestamp.v[:8])
 
         # Write size field
-        write_hdr("WRSL", "WRSH", 3, self.size.v[:8])
-        write_hdr("WRSH", "WRTL", 4, self.size.v[8:16])
+        write_hdr("WRSH", "WRSL", self.delta_ts_size.v[:3] << 5 | self.size.v[8:13])
+        write_hdr("WRSL", "WRF0", self.size.v[:8])
+
+        # Write flags field
+        write_hdr("WRF0", "WH0", self.flags.v[:8])
 
-        write_hdr("WRTL", "WRTM", 5, pkt_timestamp[:8])
-        write_hdr("WRTM", "WRTH", 6, pkt_timestamp[8:16])
-        write_hdr("WRTH", "SEND", 7, pkt_timestamp[16:24])
+        # Write header magic byte
+        write_hdr("WH0", "SEND", 0xA0)
 
         self.fsm.act("SEND",
             self.out_addr.stb.eq(1),
             self.out_addr.payload.start.eq(self.produce_header.v),
-            If(packet_too_long,
-                self.out_addr.payload.count.eq(MAX_PACKET_SIZE + 8)
-            ).Else(
-                self.out_addr.payload.count.eq(self.size.v + 8),
-            ),
+            self.out_addr.payload.count.eq(self.produce_write.v - self.produce_header.v),
             If(self.out_addr.ack,
-                self.produce_header.set(self.produce_write.v),
+                # Reserve space for (worst case) next packet header
+                self.produce_header.set(self.produce_write.v + MAX_HEADER_SIZE + 1),
```

Comment:
The `+ 1` was needed because I was getting out of sync data, but after a closer inspection, it seems that I should have modified the watermark check above to compare against `MAX_HEADER_SIZE` instead of `8`. I'll try to check it later when I get back to computer with Xilins ISE.

Reply:
Changed the watermark check and removed the `+ 1` from here as it should have been.

Under extremely heavy high-speed load I can trigger out of sync (unmatched bytes) communication, however I can also do so using the fwpkg prior to my gateware changes (so the out-of-sync root cause is unrelated to this PR).